docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,260 @@
+  }
+
+  denom = "ibc/"+ traceHash.String()  "/" + baseDenom
+  return denom
+}
+```
+
+The denomination also needs to be updated when token is received on the source chain:
+
+```golang
+// UnprefixDenom removes the first portID/channelID pair from a given denomination trace info and returns the
+// denomination with the updated trace hash and the new trace info.
+// An error is returned if the trace cannot be found on the store from the denom's trace hash.
+func (k Keeper) UnprefixDenom(ctx Context, denom string) (denom, trace string, err error) {
+  denomSplit := strings.SplitN(denom, "/", 3)
+
+  switch {
+    case denomSplit[0] == denom:
+      return "", Wrapf(ErrInvalidDenomForTransfer, "denomination should be prefixed with the format 'ibc/{traceHash}/%s'", denom)
+    case denomSplit[0] != "ibc":
+      return "", Wrapf(ErrInvalidDenomForTransfer, "denomination %s must start with 'ibc'", denom)
+  }
+
+  baseDenom = denomSplit[2]
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> trace info prefix
+  trace = k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  traceSplit := strings.SplitN(denom, "/", 3)
+  if len(traceSplit) == 3 {
+    // the trace has only one portID/channelID pair
+    return baseDenom
+  }
+
+  // remove a single identifiers pair to create the new trace
+  trace = trace[2]
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String() + "/" + baseDenom
+  return denom
+}
+```
+
+```golang
+// GetTraceFromDenom returns the token source tracing info from the trace hash of the given denomination
+func (k Keeper) GetTraceFromDenom(ctx Context, denom string) (string, error) {
+  denomSplit := strings.Split(denom, "/")
+  if denomSplit[0] == denom {
+    return "", nil
+  }
+
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> trace info prefix
+  trace := k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  return trace
+}
+
+```
+
+Additionally, the `SendTransfer`'s `createOutgoingPacket` call and the `OnRecvPacket` need to be
+updated to be retreive the trace info (using `GetTraceFromDenom`) prior to checking the correctness
+of the prefix.
+
+### Coin Changes
+
+The coin denomination validation will need to be updated to reflect these changes. In particular, the denomination validation
+function will now accept slash separators (`"/"`) and will bump the maximum character length to 64.
+
+In the specific case of cross-chain fungible token transfers, an IBC denomination will allow at
+most 27 characters for base denominations:
+
+```golang
+maxBaseDenomLen = 64 - 4 - 32 - 1  // max len - "ibc/" - SHA256 hash - "/"
+maxBaseDenomLen = 27
+```
```

Comment:
This seems odd to me. This would make it possible to deploy a coin with a denom that allows intra-chain Sends, but a denom that is too long for inter-chain transfers? This seems like something that unnecessarily adds information for developers to know. There's no need for two different length validations, for two different contexts.

I think we should just have one denom validation criteria, which if passed, will allow a coin to be sent or transferred through the interchain ecosystem.

This is possible, so long as we keep baseDenom separate. It will also allow different chains to make different decisions on maximum/minimum denom length, while keeping the IBC-relevant parts (the traceHash) consistent across every chain.

Reply:
As I said above, why can't be just truncate the base denom if it's too long? Let's not complect denom validation with IBC validation. If denoms can be 64 chars with `/`s that is the denom validation. IBC can truncate things that are too long, i.e. `"ibc/" + trace hash + "/" + first 27 chars of base denom`